Add additional sanitization for settings.typography.fontFamilies

jffng opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR adds an additional level of sanitization for theme.json's `settings.typography.fontFamilies`.

## Why?
Addresses #52798

## How?
By adding additional properties to the valid settings of the theme.json class.

## Testing Instructions
See #52798.

<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

jffng committed:
Add additional keys to valid typography.fontFamily settings.

jffng commented:
I am not sure if this is a great approach to add all of these properties to the VALID_SETTINGS array.

There is another public method of [`WP_Theme_JSON::remove_insecure_properties`](https://github.com/WordPress/gutenberg/blob/072a09dcea16d088e0f1bd49de884845d92ad777/lib/class-wp-theme-json-gutenberg.php#L2847), but when I tried using it on the example data with bad keys, it unset the entire object.

matiasbenedetto reviewed (approved):
Hi! Thanks for working on this @jffng. :) 
I tested this and it is working as expected.
LGTM

matiasbenedetto reviewed (commented):
For those testing following the instructions on https://github.com/WordPress/gutenberg/issues/52798 remember you need to create a new `WP_Theme_JSON_Gutenberg` object instead of `WP_Theme_JSON` because this change is not in core yet.

oandregal commented:
Nice one. Would you be able to provide a test case to make sure both the constructor and `remove_insecure_properties` work well?

There are a couple of examples in the existing test suite https://github.com/WordPress/gutenberg/blob/4d2a30858055a2b1d7bc04eb37a13403882391a3/phpunit/class-wp-theme-json-test.php#L847

jffng commented:
@oandregal When writing unit tests, I realized a bug about the current state of this PR — the sanitization removes any `fontFamilies` and `fontFaces` after the first one defined.

I think this is because the `remove_keys_not_in_schema` function compares the keys of the arrays and doesn't specifically account for arrays with numeric keys that represent list-like structures.

Any ideas for how to handle this?

oandregal commented:
Ah, I see, it removes all keys that are not `0`.

One approach would be to expand the logic in the [`sanitization` method](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php#L708) for the list of "valid keys" within font families to have all the indexes present in the input:

```json
{
  "...": {
    "fontFamilies": {
      "0": { "name": "...", "slug": "...", "...": "..." },
      "1": { "name": "...", "slug": "...", "...": "..." },
      "2": { "name": "...", "slug": "...", "...": "..." }
    }
  }
}
```

We do something similar for blocks, when we create the schema based on the blocks registered. The schema for the font family under the `PRESETS_METADATA`, for example (and we could add the schemas for other presets later).

A second approach that comes to mind would be to use a special sign as key to signal that "any key is allowed as child". For example, make the `VALID_SETTINGS` something like:

```json
{
  "...": {
    "fontFamilies": {
      "*": { "name": "...", "slug": "...", "...": "..." }
    }
  }
}
```

and update `sanitization` to allow any key when `*` is used in the `VALID_SETTINGS`.

Finally, another approach would be to leverage the info already present in [this other schema](https://github.com/WordPress/gutenberg/blob/trunk/schemas/json/theme.json#L551).

Does this help?

I don't feel strongly about any approach, other than making sure the way we do it can be reused for any other preset.

jffng committed:
Expand sanitization logic for fontFamilies and fontFaces.

jffng committed:
Add test for typography sanitization.

jffng committed:
Format php.

jffng commented:
Thank you for the ideas! I implemented your first suggestion here: [ba8f4de](https://github.com/WordPress/gutenberg/pull/53273/commits/ba8f4dea94c234977feaf874f245386119a85788) — it seems to be working well, the unit tests are now passing.

I also added a test to ensure the typography settings are sanitized correctly: [260d7cf](https://github.com/WordPress/gutenberg/pull/53273/commits/260d7cf7fd7f8a90d59aea080d5cf04fd4322a63)

EDIT
It looks like there are some issues with the `Tests_Fonts_WPFontsResolver_RegisterFontsFromThemeJson`, I'll look into this with @matiasbenedetto next.

matiasbenedetto commented:
> EDIT It looks like there are some issues with the `Tests_Fonts_WPFontsResolver_RegisterFontsFromThemeJson`, I'll look into this with @matiasbenedetto next.

I guess those tests were implemented by @anton-vlasenko or @hellofromtonya maybe they can help with the issues.

jffng commented:
While tying to fix the [tests](https://github.com/WordPress/gutenberg/actions/runs/5765668528/job/15632111393?pr=53273), I realized the shape of `settings.typography.fontFamilies` is different depending on the context:

With the existing fonts API, `settings.typography.fontFamilies` the input looks like:

```
[theme] => Array 
(
    [0] => Array
        (
            [fontFamily] => Lato
            [name] => Lato
            [slug] => lato
            [fontFace] => Array
                (
                    [lato-400-normal] => Array
                        (
                            [origin] => gutenberg_wp_fonts_api
                            [provider] => local
                            [fontFamily] => Lato
                            [fontStyle] => normal
                            [fontWeight] => 400
                            [fontDisplay] => fallback
                            [src] => https://example.com/tests/assets/fonts/lato/Lato-Regular.woff2
                        )
              ...
```
or

```
[theme] => Array 
(
    [0] => Array
        (
            [fontFace] => Array
                (
                    [0] => Array
                        (
                            [fontFamily] => DM Sans
                            [fontStretch] => normal
                            [fontStyle] => normal
                            [fontWeight] => 400
                            [src] => Array
                                (
                                    [0] => file:./assets/fonts/dm-sans/DMSans-Regular.woff2
                                )
                        )
...
```

With the soon-to-be introduced Font Library, the input is expected to look like this:

```
Array
(
	'fontFamily' => 'Piazzolla',
	'name'       => 'Piazzolla',
	'slug'       => 'piazzolla',
	'fontFace'   => Array
	(
		Array
		(
			'fontFamily' => 'Piazzolla',
			'fontStyle'  => 'italic',
			'fontWeight' => '400',
			'src'        => 'https://example.com/font.ttf',
		),
	),
)
```

Is there much benefit to change the sanitization function to allow all of these cases? It feels overly specific and prone to error.

Alternatively we can add sanitization directly to the new Font Library class.

oandregal commented:
That's a good discovery, glad you were looking into this.

> Is there much benefit to change the sanitization function to allow all of these cases? It feels overly specific and prone to error.

I reckon I don't know every detail of font family management, though I understand we need to consolidate into only one shape.

jffng commented:
I excluded the `fontsapi` tests per https://github.com/WordPress/gutenberg/pull/52704#issuecomment-1671855014 but there are still some failing tests in `Tests_Fonts_WPPrintFontFaces` and `Tests_Fonts_WPFontFaceResolver_GetFontsFromThemeJson`. Will these tests still be relevant when the Fonts API is removed?

hellofromtonya commented:
>I excluded the `fontsapi` tests per https://github.com/WordPress/gutenberg/pull/52704#issuecomment-1671855014 but there are still some failing tests in `Tests_Fonts_WPPrintFontFaces` and `Tests_Fonts_WPFontFaceResolver_GetFontsFromThemeJson`. Will these tests still be relevant when the Fonts API is removed?

No, these tests will not be removed. Rather, they exist to validate Font Face (the server-side `@font-face` styles generator and printer) is properly getting fonts from theme.json (via `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_settings()`) and then generating and printing the `@font-face` styles for each.

This PR is changing the structure, right? That means the font data pulled from `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_settings()` also has a different structure. And thus the changes in this PR will also modify the Font Face and its tests.

hellofromtonya commented:
Currently looking into why the Font Face tests are failing.

hellofromtonya commented:
>there are still some failing tests in `Tests_Fonts_WPPrintFontFaces` and `Tests_Fonts_WPFontFaceResolver_GetFontsFromThemeJson`.  Will these tests still be relevant when the Fonts API is removed?

@jffng Yes, these tests are still relevant and will remain. They are for Font Face, which replaces the Fonts API. It works with Font Library.

hellofromtonya commented:
Why are the Font Face tests failing?

When `WP_Font_Face_Resolver::get_fonts_from_theme_json()` is called, there is no `'fontFamilies'` in the `$settings['typography']`:

```php
public static function get_fonts_from_theme_json() {
	$settings = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_settings();

	// Bail out early if there are no font settings.
	if ( empty( $settings['typography'] ) || empty( $settings['typography']['fontFamilies'] ) ) {
		return array();
	}
```

The tests fail because it expects to receive the fonts from theme.json, but no fonts are returned (an empty array is returned).

Why is `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_settings();` not returning the `fontFamilies`?

Here's the test theme's `theme.json` https://github.com/WordPress/gutenberg/blob/trunk/phpunit/data/themedir1/fonts-block-theme/theme.json#L22.

hellofromtonya commented:
# Test Report

Env:
* Localhost: wp-env
* Theme: TT3
* Plugins: Gutenberg
* OS: macOS
* Browser: Firefox

## Testing Instructions

1. Go to Site Editor.
2. Select the Styles icon to open the Styles panel.
3. Select Typography.
4. Select Headings.

Expected behavior: The Fonts option should be there.
![Headings font before the PR](https://github.com/WordPress/gutenberg/assets/7284611/e8b546f0-141d-413c-9f2f-086ca471ac2c)

## Test Results
❌ The Fonts selector is not present.

![Fonts not present after applying the PR](https://github.com/WordPress/gutenberg/assets/7284611/9bd9c7f7-429e-459c-b8b4-f6e327c4ce62)

@jffng after applying the changes in this PR, the Fonts option and dropdown is no longer present in Site Editor > Styles > Typography > Headings UI.

jffng committed:
Fix case where a source is set.

jffng committed:
Format php.

jffng commented:
I updated this PR so all the tests are now passing and the above bug reports should no longer be happening, it should be ready for another review.

ironprogrammer commented:
This works as advertised, and local PHP unit tests pass 👍🏻 

## Test Report
Includes test from @hellofromtonya's [report above](https://github.com/WordPress/gutenberg/pull/53273#issuecomment-1681298646), as well as the direct sanitation check from https://github.com/WordPress/gutenberg/issues/52798.

### Steps to Test
1. UI: Check for the font picker under *Styles > Typography > Headings > Typography*.
2. Plugin: Check sanitation directly by installing [this mu-plugin](https://gist.github.com/ironprogrammer/8aa37a77cf3989efaadd32f1fa7028ee). Reload the site to view result. _Remove plugin after testing._

### Environment
- OS: macOS 13.5.2
- Browser: Safari 16.6
- Server: nginx/1.25.2
- PHP: 7.4.33
- WordPress: 6.4-alpha-56267-src
- Theme: twentytwentythree v1.2
- Active Plugins:
  - gutenberg v16.6.0 (this PR)
  - test-gb-52798 ([test mu-plugin](https://gist.github.com/ironprogrammer/8aa37a77cf3989efaadd32f1fa7028ee))

### Actual Results
- ✅ UI: The font dropdown is visible/functional.
- ✅ Plugin: The invalid values were removed from the final theme data using the patched `WP_Theme_JSON_Gutenberg` class.

<img width="733" alt="sanitized theme data" src="https://github.com/WordPress/gutenberg/assets/824344/965baca9-24f9-4efa-a67e-f32715f84b4e">

jffng commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -796,6 +819,23 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
+		// For settings.typography.fontFamilies, make the $schema have all indexes present in the $input.
+		if ( isset( $input['settings']['typography']['fontFamilies'] ) ) {
+			// Do not handle the cases where the sanitization is called before font families are merged, since it will be handled again later.
+			if ( isset( $input['settings']['typography']['fontFamilies']['theme'] ) || isset( $input['settings']['typography']['fontFamilies']['custom'] ) ) {
```
There are some cases where `settings.typography.fontFamilies` have a "source" which is either `theme` or `custom`. I thought this was the simplest way to handle this case because it breaks the validation otherwise. But I'm not sure if it exposes an insecurity since we're returning early.

jffng replied:
Addressed this with @vcanales in [38263a1](https://github.com/WordPress/gutenberg/pull/53273/commits/38263a16151cc181a869bee893653b679c49ba54).

mikachan commented:
As this PR hasn't been merged yet and the latest version of Gutenberg is due to be released today, I'm going to remove the `Backport to Gutenberg RC` label.

I'm going to add the `Backport to WP Beta/RC` label as the Font Library work is a "blessed" task for WordPress 6.4.

jffng committed:
Fix bug in flattening function and expand test case.

Commit:
Run phpcbf on changes

hellofromtonya commented:
The Font Face tests are failing because they are using Core's Font Face files which do not use `WP_Theme_JSON_Gutenberg`. 

Seems Font Face files in Gutenberg need to be modified to use `_Gutenberg` for running GB specific code. I'll follow-up with a PR shortly to fix the failing tests.

Note: Once this PR is merged into Core, the tests will pass again. But the problem will resurface anytime there's a change in one of the files here in this repo that already exists in Core.

hellofromtonya commented:
PR https://github.com/WordPress/gutenberg/pull/54990 should resolve the failing tests. @ironprogrammer can you test to verify please?

jffng committed:
Handle case where fontFamilies have an origin e.g. theme or custom.

matiasbenedetto committed:
Merge branch 'trunk' into fix/theme-json-fonts-validation

jffng commented:
Thanks everyone for the input, it helped inform a cleaner solution at https://github.com/WordPress/gutenberg/pull/56447.
